actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,77 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }
+      rendered_partials = @collection.any? ? yield : []
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      if ActionController::Base.perform_caching
+        @options[:cache].present? || automatic_cache_eligable?
+      end
+    end
+
+    def automatic_cache_eligable?
+      if collection_only_renders_one_template? && !custom_collection_cache_key?
+        @template.source =~ /\A<% cache\(?\s*(\w+)/ && $1 == inferred_cache_name.to_s
+      end
+    end
+
+    def collection_only_renders_one_template?
+      @template # Template is only set when a collection renders one template.
```

Comment:
@dhh When you render a collection, do you render have it render different templates? Say a todo list renders `todo` and `amazing_todo`?

If so, we can't use `read_multi` there.

Because `@template` is [only set](https://github.com/rails/rails/blob/master/actionview/lib/action_view/renderer/partial_renderer.rb#L294) if there's a `@path` and `@path` is only set if the collection renders [a single template](https://github.com/rails/rails/blob/master/actionview/lib/action_view/renderer/partial_renderer.rb#L378-L380).

Eventually `collection_without_template` is called and we don't have access to where [it finds templates](https://github.com/rails/rails/blob/master/actionview/lib/action_view/renderer/partial_renderer.rb#L458).

We could try and patch it in there, but then we'd fetch the templates twice. I'll admit I don't know how costly that is. But I'd rather not :smile: